elftools/elf/relocation.py

```diff
@@ -296,18 +297,17 @@ def _do_apply_relocation(self, stream, reloc, symtab):
     # Relocations are represented by "recipes". Each recipe specifies:
     #  bytesize: The number of bytes to read (and write back) to the section.
     #            This is the unit of data on which relocation is performed.
-    #  has_addend: Does this relocation have an extra addend?
```

Comment:
Sure. I removed `has_addend` because it was the easiest way I found to make what I needed work.

See commit ec3d215a29c6f531a37c47c1760786261067fd07 which is before removing `has_addend`. It has this test failing:

```
❯ venv/bin/python scripts/readelf.py --debug-dump=info test/testfiles_for_readelf/simple_mips_gcc.o.elf
Traceback (most recent call last):
  File "scripts/readelf.py", line 1852, in <module>
    main()
  File "scripts/readelf.py", line 1828, in main
    readelf.display_debug_dump(args.debug_dump_what)
  File "scripts/readelf.py", line 899, in display_debug_dump
    self._init_dwarfinfo()
  File "scripts/readelf.py", line 1104, in _init_dwarfinfo
    self._dwarfinfo = self.elffile.get_dwarf_info()
  File "./elftools/elf/elffile.py", line 277, in get_dwarf_info
    relocate_dwarf_sections)
  File "./elftools/elf/elffile.py", line 795, in _read_dwarf_section
    section_stream, reloc_section)
  File "./elftools/elf/relocation.py", line 216, in apply_section_relocations
    self._do_apply_relocation(stream, reloc, symtab)
  File "./elftools/elf/relocation.py", line 287, in _do_apply_relocation
    addend=reloc['r_addend'] if recipe.has_addend else 0)
  File "./elftools/elf/relocation.py", line 39, in __getitem__
    return self.entry[name]
  File "./elftools/construct/lib/container.py", line 35, in __getitem__
    return self.__dict__[name]
KeyError: 'r_addend'
```

This is because I changed `ENUM_RELOC_TYPE_MIPS['R_MIPS_32']` to have `has_addend=True`. If I change it back to `False`, this fails:

```
❯ venv/bin/python test/run_dwarfdump_tests.py test/testfiles_for_dwarfdump/dwarf_mips64el.o.elf
Test file 'test/testfiles_for_dwarfdump/dwarf_mips64el.o.elf'
.......................FAIL
....for file test/testfiles_for_dwarfdump/dwarf_mips64el.o.elf
....for option "--debug-info"
....Output #1 is llvm-dwarfdump, Output #2 is pyelftools
@@ Mismatch on line #6:
>>              dw_at_name [dw_form_strp]	( .debug_str[0x000000dc] = "./dwarf_mips64el.c")<<
>>              dw_at_name [dw_form_strp]	( .debug_str[0x00000000] = "gnu c17 12.2.0 -mel -march=mips64r2 -mabi=64 -mllsc -mips64r2 -g -o2 -fpic -fstack-protector-strong -fno-strict-overflow -frandom-seed=kdyy43gwzl --param=ssp-buffer-size=4")<<
 ([('equal', 0, 61, 0, 61), ('replace', 61, 63, 61, 63), ('equal', 63, 68, 63, 68), ('insert', 68, 68, 68, 78), ('equal', 68, 69, 78, 79), ('replace', 69, 76, 79, 95), ('equal', 76, 82, 95, 101), ('replace', 82, 86, 101, 239), ('equal', 86, 88, 239, 241)])
@@ Output #1 dumped to file: /tmp/out-1-dwarf_mips64el.o.elf---debug-info-l4hdubyk.stdout
@@ Output #2 dumped to file: /tmp/out-2-dwarf_mips64el.o.elf---debug-info-ng5agd0v.stdout

Conclusion: FAIL
```

I have figured it was OK to remove this field from the recipe, since you can know if there is an `r_addend` field from the relocation object.

Anyway, I now created another branch without removing the `r_addend` field, which instead uses two recipe dicts instead of one: `_RELOCATION_RECIPES_MIPS_RELA` and `_RELOCATION_RECIPES_MIPS_REL`. It's here: https://github.com/noamraph/pyelftools/tree/mips64-o-with-has-addend

What do you think? Do you prefer the latter approach?

Thanks!
Noam

Reply:
Could you please send your other branch as a separate PR? It should be easy to create a new PR from a different branch. A PR lets me see the delta/patch of the change much more conveniently than looking at a whole branch

Reply:
Sure. Here it is: https://github.com/eliben/pyelftools/pull/495